#166789997 Users Should Be Able to Create Articles

Hervera opened the pull request:
#### What does this PR do?
It creates article with all CRUD operation

#### Description of Task to be completed
- create article
- retrieve all articles
- retrieve single article
- update article
- delete article

#### How should this be manually tested?
1. Run the following commands in your terminal

```
git clone https://github.com/andela/ah-92explorers-backend.git
cd ah-92explorers-backend
git checkout ft-create-articles-166789997
npm install
npm start
```

2. Go to Postman

6. Run the following requests in Postman using `localhost:3000`
## Endpoints
|  Method  |  Endpoint  |  Task  | Requirement |
|  --- |  --- |  ---  | ---  |
|  `POST`  |  `/api/users`  |  `Register`  | `No Authorization`|
|  `POST`  |  `/api/users/login`  |  `Login`  | `No Authorization`|
|  `POST`  |  `/api/articles`  |  `Login`  | `Authorization with a valid token`|
|  `GET`  |  `/api/articles`  |  `fetches all articles`  | `No Authorization`|
|  `GET`  |  `/api/article/:slug`  |  `fetches a single article`  | `No Authorization`|
|  `PUT`  |  `/api/article/:slug`  |  `updates an article`  | `Authorization with ownership`|
|  `DELETE`  |  `/api/article/:slug`  |  `deletes an article`  | `Authorization with ownership`|

#### Any background context you want to provide?
- N/A

#### What are the relevant pivotal tracker stories?
[#166789997](https://www.pivotaltracker.com/story/show/166789997)

#### Screenshots (if appropriate)
`Insert article`
<img width="1089" alt="Screen Shot 2019-07-15 at 12 05 13" src="https://user-images.githubusercontent.com/25223709/61224637-36febe00-a71f-11e9-9563-13fd6b6edd5c.png">

`Retrieve all articles`
<img width="1089" alt="Screen Shot 2019-07-15 at 12 05 57" src="https://user-images.githubusercontent.com/25223709/61224675-4a118e00-a71f-11e9-8f51-367c8f0ca28e.png">

`Retrieve single article`
<img width="1089" alt="Screen Shot 2019-07-15 at 12 06 21" src="https://user-images.githubusercontent.com/25223709/61224695-54338c80-a71f-11e9-86a8-ce898ddd38b8.png">

`Update article`
<img width="1089" alt="Screen Shot 2019-07-15 at 12 17 04" src="https://user-images.githubusercontent.com/25223709/61224724-631a3f00-a71f-11e9-8710-5600a78bf62d.png">

`Delete article`
<img width="1089" alt="Screen Shot 2019-07-15 at 07 03 07" src="https://user-images.githubusercontent.com/25223709/61196403-b0ba8b80-a6ce-11e9-957a-3741a08ac38e.png">


#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I don't have more than 3 major commits on this PR

niyoceles reviewed (changes requested):
squash the commits in one commit

Commit:
feature(create-articles): CRUD operations for articles

- create migration
- create article model
- add relationship between users and articles tables
- add crud functionalities
- add routes
- add tests

[Start #166789997]

elemanhillary-zz reviewed (changes requested):
resolve the merge conflict

elemanhillary-zz commented on controllers/articleController.js:
```diff
@@ -0,0 +1,153 @@
+        include: [
+          {
+            model: users,
+            as: 'authorfkey',
```
this will return `authorfkey` instead of `author`. kindly change that so that it reflects the meaning of the return key

kodek-sleuth commented on controllers/article.js:
```diff
@@ -0,0 +1,154 @@
+import models from '../models';
+
+const { articles, users } = models;
+const clientUrl = (process.env.NODE_ENV === 'production') ? 'https://ah-92explorers-api.herokuapp.com' : 'http://127.0.0.1:3000';
```
This is an unnecessary line

kodek-sleuth commented on controllers/article.js:
```diff
@@ -0,0 +1,154 @@
+/**
+ * @param {class} --Article controller
+ */
+class ArticleController {
```
Rename this to Article

kodek-sleuth commented on controllers/article.js:
```diff
@@ -0,0 +1,154 @@
+   * @param {Object} res  get all created articles
+   * @returns {Object} return all created articles
+   */
+  static async index(req, res) {
```
Please rename this and follow more descriptive naming

kodek-sleuth commented on tests/dummy.js:
```diff
@@ -0,0 +1,9 @@
+const article1 = {
```
Please use seeders and this code should be in the index.test.js file

kodek-sleuth commented on tests/article.test.js:
```diff
@@ -0,0 +1,193 @@
+  });
+  // @delete article
+  // return status code of 404
+  it('should not delete an article which doesn\'t', (done) => {
```
Please be more descriptive in this it block

kodek-sleuth commented on tests/article.test.js:
```diff
@@ -0,0 +1,193 @@
+  });
+  // @delete article
+  // @return status code of 204
+  it('should delete an article return status code of 204 on deleting article', (done) => {
```
Please correct this grammar

kodek-sleuth commented on tests/article.test.js:
```diff
@@ -0,0 +1,193 @@
+      });
+  });
+  // @get single articles
+  it('Should get single article ', (done) => {
```
Correct this grammar

kodek-sleuth commented on tests/article.test.js:
```diff
@@ -0,0 +1,193 @@
+  });
+
+  // @create article
+  it('Should create article ', (done) => {
```
Grammar

kodek-sleuth commented on tests/article.test.js:
```diff
@@ -0,0 +1,193 @@
+let articleSlug;
+
+describe('CRUD articles Routes', () => {
+  before(() => {
```
To further help you on the matter why seeders are important, in this case, you have to run two blocks of code in order to carry out the main test. You could have used seeders so that you just login the seeded user

kodek-sleuth commented on middlewares/article.js:
```diff
@@ -0,0 +1,24 @@
+*/
+export const checkingArticle = async (req, res, next) => {
+  try {
+    const findArticle = await articles.findOne({ where: { slug: req.params.slug } });
```
Destructure your objects

Hervera replied:
rename `findArticle` to `article`

elemanhillary-zz commented on tests/dummy.js:
```diff
@@ -0,0 +1,9 @@
+const article1 = {
```
we already have a file which holds all dummy data, please make use of that

kodek-sleuth replied:
Please also seed article data in the db that will be used by other peers as they work on this object

kodek-sleuth commented on tests/article.test.js:
```diff
@@ -0,0 +1,193 @@
+  });
+  // @delete article
+  // return status code of 404
+  it('should not delete an article which doesn\'t found', (done) => {
```
please check your grammar as you create your code

elemanhillary-zz commented on controllers/article.js:
```diff
@@ -0,0 +1,162 @@
+      });
+      return res.status(200).json({ articles: allArticles });
+    } catch (error) {
+      return res.status(500).json(`${error}`);
```
dont give a user all error stacktrace, replace with a discreptive error message

kodek-sleuth reviewed (dismissed):
This PR is ready to be merged

IsaiahRn reviewed (dismissed):
This PR is ready to be merged!

elemanhillary-zz reviewed (dismissed):
ready for merge

Hervera commented on controllers/article.js:
```diff
@@ -0,0 +1,162 @@
+    try {
+      // @retrieve articles
+      const allArticles = await articles.findAll({
+        order: [['id', 'DESC']],
```
sort by `createdAt`

Hervera commented on middlewares/article.js:
```diff
@@ -0,0 +1,24 @@
+export const checkingArticle = async (req, res, next) => {
+  try {
+    const findArticle = await articles.findOne({ where: { slug: req.params.slug } });
+    const findUser = await users.findOne({ where: { email: req.decoded.email } });
```
rename `findUser` to `author`

Hervera commented on tests/article.test.js:
```diff
@@ -0,0 +1,171 @@
+      .attach('image', '')
+      .end((error, res) => {
+        if (error) done(error);
+        res.should.have.status(200);
```
Check the message in the response

elemanhillary-zz reviewed (dismissed):
ready for merge

elemanhillary-zz reviewed (approved):
ready for merge
